skrub/_gap_encoder.py

```diff
@@ -251,13 +306,25 @@ def fit(self, X: ArrayLike, y=None) -> "GapEncoderColumn":
                     self.rescale_W,
                     self.rho_,
                 )
-
-                if i == n_batch - 1:
-                    # Compute the norm of the update of W in the last batch
-                    W_change = np.linalg.norm(self.W_ - W_last) / np.linalg.norm(W_last)
-
-            if (W_change < self.tol) and (n_iter_ >= self.min_iter - 1):
-                break  # Stop if the change in W is smaller than the tolerance
+                batch_cost = _beta_divergence(
+                    unq_V[idx],
+                    unq_H[idx],
+                    self.W_,
+                    "kullback-leibler",
+                    square_root=False,
+                ) / len(idx)
+                if self._minibatch_convergence(
+                    batch_size=len(idx),
+                    batch_cost=batch_cost,
+                    n_samples=n_samples,
+                    step=n_iter_ * n_batch + i,
+                    n_steps=self.max_iter * n_batch,
+                ):
+                    break
+            else:
+                # only continue if no break occurred
+                continue
+            break
```

Comment:
This is pretty hard to read. Should we put the second for-loop in a function and use `return` instead? Alternatively, we could use a boolean flag, but the current function will look cluttered. WDYT?

Reply:
I like the boolean solution, it's easier to read